cdap-authorization-ldap-role/src/main/java/io/cdap/cdap/security/authorization/ldap/role/searcher/LDAPSearcher.java

```diff
@@ -0,0 +1,147 @@
+  private static final int MAX_RETRIES = 5;
+
+  private final LDAPSearchConfig config;
+  private final String[] baseDNList;
+
+  public LDAPSearcher(LDAPSearchConfig config) {
+    this.config = config;
+    baseDNList = config.getSearchBaseDn().split(BASE_DN_SPLITTER);
+  }
+
+  public void testConnection() {
+    try {
+      DirContext context = getConnection();
+      context.close();
+    } catch (NamingException e) {
+      LOG.error("Failed to establish connection to '{}'", config.getUrl(), e);
+      throw new RuntimeException(e);
+    }
+  }
+
+  public Set<String> searchGroups(String username) {
+    Set<String> groups = Collections.emptySet();
+
+    try {
+      DirContext context = getConnection();
+
+      SearchControls controls = new SearchControls();
+      if (config.isRecursiveSearch()) {
+        controls.setSearchScope(SUBTREE_SCOPE);
+      }
+
+      groups = Arrays.stream(baseDNList)
+        .map(baseDN -> searchGroups(baseDN, username, context, controls))
+        .flatMap(Collection::stream)
+        .collect(Collectors.toSet());
+
+      context.close();
+    } catch (NamingException e) {
+      LOG.error("Failed to find groups for '{}'", username, e);
+    }
+
+    return groups;
+  }
+
+  private Set<String> searchGroups(String baseDN, String username, DirContext context, SearchControls controls) {
+    String filter = String.format(config.getSearchFilter(), username);
+    Set<String> groups = new HashSet<>();
+
+    try {
+      NamingEnumeration<SearchResult> renum = context.search(baseDN, filter, controls);
+
+      if (!renum.hasMore()) {
+        LOG.warn("Cannot locate user information for '{}'", username);
+        return groups;
+      }
+
+      SearchResult result = renum.next();
+
+      Attribute memberOf = result.getAttributes().get(config.getMemberAttribute());
+      if (memberOf != null) {
+        for (int i = 0; i < memberOf.size(); i++) {
+          groups.add(memberOf.get(i).toString());
+        }
+      }
+    } catch (PartialResultException e) {
+      LOG.warn("Failed to find groups for '{}' in '{}'", username, baseDN);
+    } catch (NamingException e) {
+      LOG.error("Failed to find groups for '{}' in '{}'", username, baseDN, e);
+    }
+
+    return groups;
+  }
+
+  private DirContext getConnection() throws NamingException {
+    Hashtable<String, String> props = new Hashtable<>();
+    String url = config.getUrl();
+
+    props.put(Context.SECURITY_PRINCIPAL, config.getLookUpBindDN());
+    props.put(Context.SECURITY_CREDENTIALS, config.getLookUpBindPassword());
+
+    if (config.isIgnoreSSLVerify() && url.startsWith(LDAPS_PROTOCOl)) {
+      props.put(LDAP_SOCKET_FACTORY, BlindSSLSocketFactory.class.getName());
+    }
+
+    NamingException exception = new NamingException();
+    for (int i = 1; i <= MAX_RETRIES; i++) {
+      try {
+        return LdapCtxFactory.getLdapCtxInstance(url, props);
+      } catch (NamingException e) {
```

Comment:
it could be other retriable exceptions for sure (at least IOExceptions)

Reply:
LdapCtxFactory.getLdapCtxInstance throws only NamingException checked exception, I suppose that it wraps other exceptions in NamingException